module/apmawssdkgo/session.go

```diff
@@ -66,13 +72,49 @@ type service interface {
 	setAdditional(*apm.Span)
 }
 
+func build(req *request.Request) {
+	spanSubtype := req.ClientInfo.ServiceName
+	spanType, ok := serviceTypeMap[spanSubtype]
+	if !ok {
+		return
+	}
+
+	if spanSubtype == serviceSQS && !supportedSQSMethod(req) {
+		// Unsupported SQS method type.
+		return
+	}
+
+	ctx := req.Context()
+	tx := apm.TransactionFromContext(ctx)
+	if tx == nil {
+		return
+	}
+
+	// The span name is added in the `send()` function, after other
+	// handlers have generated the necessary information on the request.
+	span := tx.StartSpan("", spanType, apm.SpanFromContext(ctx))
+	if !span.Dropped() {
+		ctx = apm.ContextWithSpan(ctx, span)
+		defer req.SetContext(ctx)
+	} else {
+		span.End()
+		span = nil
+		return
+	}
+
+	if req.ClientInfo.ServiceName != serviceSQS {
+		return
+	}
+	addMessageAttributes(req, span)
+}
+
 func send(req *request.Request) {
```

Comment:
should we moving all the logic in send over to build?

Reply:
Later steps construct the URL which we use to create the span name, for example (bucket, queue name, etc are pulled from the URL), and this isn't available until after the request has been built. However, we need a hook before the request is built to inject our own message attributes.

--> our build (add additional tracing data)
--> aws build (request generated)
--> our send (grab data from url)
--> aws send
--> our complete

Reply:
thanks for the explanation. might be worth adding a comment somewhere